File: part3_research_agent/part3/flow.py
"""Part 3 Research Agent: full flow from user query to Part3Output."""

from __future__ import annotations

from part3.types import PaperResult, Part3Output
from part3.nemotron import nemotron_plan, nemotron_rank
from part3.sources import search_semantic_scholar, search_arxiv, search_openalex


def run_research_agent(
    user_query: str,
    *,
    api_key: str,
    max_papers_per_source: int = 10,
    top_k: int = 5,
) -> Part3Output:
    """Run the research paper flow: plan -> search -> rank -> output."""
    search_queries_used: list[str] = []

    # Step 1: Nemotron plan
    search_query, preferred_sources = nemotron_plan(user_query, api_key)
    search_query = (search_query or user_query).strip()
    search_queries_used.append(search_query)

    # Step 2: Call sources
    all_candidates: list[PaperResult] = []
    sources = [s.lower() for s in preferred_sources]

    if "semantic_scholar" in sources:
        try:
            all_candidates.extend(
                search_semantic_scholar(search_query, limit=max_papers_per_source)
            )
        except Exception:
            pass

    if "arxiv" in sources:
        try:
            all_candidates.extend(
                search_arxiv(search_query, max_results=max_papers_per_source)
            )
        except Exception:
            pass

    if "openalex" in sources:
        try:
            all_candidates.extend(
                search_openalex(search_query, per_page=max_papers_per_source)
            )
        except Exception:
            pass

    if not all_candidates:
        return Part3Output(
            papers=[],
            search_queries_used=search_queries_used,
            summary="No papers found. Try a different query or check your connection.",
        )

    # Step 3: Nemotron rank
    candidates_as_dicts = [
        {
            "title": c.get("title"),
            "source": c.get("source"),
            "paper_id": c.get("paper_id"),
            "abstract": c.get("abstract"),
        }
        for c in all_candidates
    ]
    selected_indices, summary = nemotron_rank(
        user_query, candidates_as_dicts, api_key, top_k=top_k
    )

    # Step 4: Build output
    papers = [all_candidates[i] for i in selected_indices if 0 <= i < len(all_candidates)]

    return Part3Output(
        papers=papers,
        search_queries_used=search_queries_used,
        summary=summary,
    )

File: part2-dataset-agent/lib/flow.ts
/**
 * Part 2 Dataset Agent: full flow from user input to Part2Output.
 */

import type { DatasetCandidate } from "./sources/kaggle";
import type { Part2Dataset, Part2Output } from "./types";
import { nemotronPlan, nemotronRank } from "./nemotron";
import { searchKaggle } from "./sources/kaggle";
import { searchGoogleDatasetSearch } from "./sources/googleDatasetSearch";

export interface FlowEnv {
  NIM_API_KEY: string;
  KAGGLE_USERNAME?: string;
  KAGGLE_KEY?: string;
  GOOGLE_API_KEY?: string;
  GOOGLE_CSE_ID?: string;
}

export async function runDatasetAgent(
  userInput: string,
  env: FlowEnv
): Promise<Part2Output> {
  const search_queries_used: string[] = [];

  // Step 1: Nemotron plan
  const plan = await nemotronPlan(userInput, env.NIM_API_KEY);
  const query = plan.search_query.trim() || userInput;
  search_queries_used.push(query);

  // Step 2: Call sources
  const sources = plan.preferred_sources.map((s) => s.toLowerCase());
  const allCandidates: DatasetCandidate[] = [];

  if (sources.includes("kaggle") && env.KAGGLE_USERNAME && env.KAGGLE_KEY) {
    try {
      const kaggle = await searchKaggle(query, {
        username: env.KAGGLE_USERNAME,
        key: env.KAGGLE_KEY,
        maxResults: 10,
      });
      allCandidates.push(...kaggle);
    } catch (e) {
      console.error("Kaggle search failed:", e);
    }
  }

  if (sources.includes("google_dataset_search")) {
    const google = await searchGoogleDatasetSearch(query, {
      apiKey: env.GOOGLE_API_KEY,
      cseId: env.GOOGLE_CSE_ID,
      maxResults: 5,
    });
    allCandidates.push(...google);
  }

  if (allCandidates.length === 0) {
    return {
      datasets: [],
      search_queries_used,
      summary: "No datasets found. Add Kaggle credentials (KAGGLE_USERNAME, KAGGLE_KEY) or Google Dataset Search (GOOGLE_API_KEY, GOOGLE_CSE_ID).",
    };
  }

  // Step 3: Nemotron rank
  const { selected_slugs, summary } = await nemotronRank(
    userInput,
    allCandidates,
    env.NIM_API_KEY
  );

  // Step 4 & 5: Build output (we return metadata + URLs; actual download can be done by Part 4 or frontend)
  const byKey = (c: DatasetCandidate) => `${c.source}:${c.slug_or_id}`;
  const selectedSet = new Set(selected_slugs.map((s) => `${s.source}:${s.slug_or_id}`));
  const selected = allCandidates.filter((c) => selectedSet.has(byKey(c)));

  const datasets: Part2Dataset[] = selected.map((c) => ({
    source: c.source,
    name: c.name,
    slug_or_id: c.slug_or_id,
    path_or_url: c.url || c.slug_or_id,
    format: (c.format as Part2Dataset["format"]) || "unknown",
    description: c.description,
    size: c.size,
  }));

  return {
    datasets,
    search_queries_used,
    summary,
  };
}
